### PR TITLE
Add verify committed offset test

### PR DESCRIPTION
**Pull request by pierDipi**
Verify that after sacura run, the consumer group lag
`Lag = Sum(offset to produce next - committed offset)`
is 0.

While I was creating the test, I've seen an opportunity to
create a reusable (and testable in isolation) library to
calculate consumer group lags.

## Proposed Changes

- Add verify committed offset test

### CODE REVIEW

**Comment by pierDipi** [comment]
Maven
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.2.4:shade (default) on project receiver: Execution default of goal org.apache.maven.plugins:maven-shade-plugin:3.2.4:shade failed: Plugin org.apache.maven.plugins:maven-shade-plugin:3.2.4 or one of its dependencies could not be resolved: Could not transfer artifact org.ow2.asm:asm:jar:8.0 from/to central (https://repo.maven.apache.org/maven2): Transfer failed for https://repo.maven.apache.org/maven2/org/ow2/asm/asm/8.0/asm-8.0.jar: Connection reset -> [Help 1]
```
/test pull-knative-sandbox-eventing-kafka-broker-unit-tests

**Comment by slinkydeveloper** [comment]
What's the lag?

**Comment by pierDipi** [comment]
`Lag = Sum(offset to produce next - committed offset)`

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (slinkydeveloper) [inline control-plane/pkg/kafka/consumer_group_lag.go:49]: Maybe rename `PartitionOffsetStatus`?

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (slinkydeveloper) [inline control-plane/pkg/kafka/consumer_group_lag.go:52]: Don't store it, it's more clear if it's not store IMO

File: test/e2e/consumer_group_lag_provider_test.go
Comment (slinkydeveloper) [inline test/e2e/consumer_group_lag_provider_test.go:33]: Can you use the new test framework for this e2e test? I would love to avoid adding new tests using the old test code, if possible

**Review by slinkydeveloper — COMMENTED** [review]
TBH the lag name just doesn't sound right to me, except that I'm ok with this PR and thanks for tackling it!

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:49]: It's pretty common in the Kafka world `Lag`, in fact, this command:
```./bin/kafka-consumer-groups.sh --bootstrap-server localhost:9092 --group mygroup  --describe```
produces an output like this:
```
TOPIC                          PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG        CONSUMER-ID                                       HOST
```

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:52]: Sure!

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:49]: Lag in the output is exactly `PartitionLag` in our code.

File: test/e2e/consumer_group_lag_provider_test.go
Comment (pierDipi) [reply test/e2e/consumer_group_lag_provider_test.go:33]: Great suggestion! Let me check what I can do with it.

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (slinkydeveloper) [reply control-plane/pkg/kafka/consumer_group_lag.go:49]: Ok I wasn't aware of that, i'm sorry, then lag it is

Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:49]: No problem!

File: test/e2e/consumer_group_lag_provider_test.go
Comment (pierDipi) [reply test/e2e/consumer_group_lag_provider_test.go:33]: I tried some stuff, but this requires some other changes to some of our helpers, that don't belong to this PR.
Also, I'm a little worried about the flakiness we've seen with the new test framework in eventing.
With that being said, I'd like to wait for this migration for now.

Comment (slinkydeveloper) [reply test/e2e/consumer_group_lag_provider_test.go:33]: Ok, then we'll do that in another PR

**Comment by slinkydeveloper** [comment]
Can you fix the lint?

**Comment by pierDipi** [comment]
Sure.

/hold

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (slinkydeveloper) [inline control-plane/pkg/kafka/consumer_group_lag.go:149]: This offset strategy thing is weird... I don't see why I should provide it to default some value. If no offset is committed, the lag is just 0 and the consumerOffset is -1, right?

**Comment by slinkydeveloper** [comment]
More than having more test images to run some specific asserts on kafka broker, can we run them in the test directly accessing to the kafka cluster?

Exposing the kafka cluster should be straightforward with strimzi: https://strimzi.io/docs/operators/latest/using.html#assembly-configuring-external-listeners-str just use a nodeport svc?

File: control-plane/pkg/kafka/consumer_group_lag.go
Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:149]: No committed offset doesn't mean that the lag is 0 because the producer might have produced messages and depending on the strategy it could be 0 (OffsetNewest) or N (OffsetOldest).

Comment (slinkydeveloper) [reply control-plane/pkg/kafka/consumer_group_lag.go:149]: Then can we just represent the state "no offset committed"? Otherwise, this information get lost, which might be useful otherwise (e.g. I'm thinking to unordered delivery tests)

Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:149]: I think propagating "no offset committed" to the caller is a good idea but I feel the logic to determine the consumer offset naturally belongs to this function.

I'd add a field called `OffsetCommitted` to determine whether an offset has been committed.

WDYT?

Comment (slinkydeveloper) [reply control-plane/pkg/kafka/consumer_group_lag.go:149]: Ok

Comment (pierDipi) [reply control-plane/pkg/kafka/consumer_group_lag.go:160]: Done

**Comment by pierDipi** [comment]
> Exposing the kafka cluster should be straightforward with strimzi: https://strimzi.io/docs/operators/latest/using.html#assembly-configuring-external-listeners-str just use a nodeport svc and then get the url retrieving the svc?

NodePort svcs require special network setups in some environments (firewall rules, etc) to be accessed from the outside, I'd prefer to keep these checks in Jobs running inside the cluster.

**Comment by slinkydeveloper** [comment]
> NodePort svcs require special network setups in some environments (firewall rules, etc) to be accessed from the outside, I'd prefer to keep these checks in Jobs running inside the cluster.

IMO I think this is worth the effort: there are other tests I would love to develop that directly connects to kafka, so we should at least give it a try... I'll try to draft something and I'll let you know.

File: test/test_images/committed-offset/admin.go
Comment (slinkydeveloper) [inline test/test_images/committed-offset/admin.go:56]: can you print the lag data structure as json, parse it in the test runner and then assert there? I prefer to use these test images only as "data collectors", so we can more flexibly perform assertions and debug in the test runner.

File: test/test_images/consumer-group-lag-provider-test/admin.go
Comment (slinkydeveloper) [inline test/test_images/consumer-group-lag-provider-test/admin.go:31]: Is this thing testing your test code? If that's the case, do we really need it? Your code to check the consumer group lag is quite easy, unit test should be enough i guess?

**Review by slinkydeveloper — COMMENTED** [review]
I tried to go down the path to expose kafka, it's way too complicated and makes worse the dev experience (which is already quite convoluted today)

So let's go down the test_images road. I've left a comment about the test_images approach, in order to make a bit easier debugging etc

File: test/test_images/committed-offset/admin.go
Comment (pierDipi) [reply test/test_images/committed-offset/admin.go:56]: Sounds interesting but parsing logs seems less robust.

I'm not sure how we use k8s' events for this kind of communication but maybe we can use a similar pattern?

File: test/test_images/consumer-group-lag-provider-test/admin.go
Comment (pierDipi) [reply test/test_images/consumer-group-lag-provider-test/admin.go:31]: I prefer to keep this test, I discovered a bug with this test :upside_down_face: which is a signal that it's useful.
I was debugging the data plane but the bug was actually in the test code.

Comment (slinkydeveloper) [reply test/test_images/consumer-group-lag-provider-test/admin.go:31]: I see... TBH, an integration test for the test code is weird :smile: Maybe can you just comment that this is useful for testing the test code and maybe disable it in our test pipeline?

File: test/test_images/committed-offset/admin.go
Comment (slinkydeveloper) [reply test/test_images/committed-offset/admin.go:56]: > I'm not sure how we use k8s' events for this kind of communication but maybe we can use a similar pattern?

Way too complicated, I suggest we avoid going down that road. WDYM by parsing logs is less robust?

File: test/test_images/consumer-group-lag-provider-test/admin.go
Comment (pierDipi) [reply test/test_images/consumer-group-lag-provider-test/admin.go:31]: Disabled.

File: test/test_images/committed-offset/admin.go
Comment (pierDipi) [reply test/test_images/committed-offset/admin.go:56]: Something like
```
	log.Printf(`{"lag": %d}`, lag.Total())
```

and then in the test should I try to parse every log line looking for the above Json? 

Is this what you mean? It seems weird. :thinking:

Comment (slinkydeveloper) [reply test/test_images/committed-offset/admin.go:56]: How about just parse the last log line? Is that reasonable?

**Review by slinkydeveloper — APPROVED** [review]
@pierDipi can we merge this one? If yes, remove the hold.

/lgtm

**Pull request merged**
